File: c++/54_Spiral_Matrix.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int> >& matrix) {
    	vector<int> res;
    	int m = matrix.size();
    	if (m == 0)
    		return res;
    	int n = matrix[0].size();
    	int direction = 0;
    	int left = 0, right = n - 1;
    	int up = 0, down = m - 1;
    	while(true) {
    		if (direction == 0) { // up lane, go right
    			for (int i = left; i <= right; i++)
    				res.push_back(matrix[up][i]);
    			up++;
    		}
    		if (direction == 1) { // right lane, go down
    			for (int i = up; i <= down; i++)
    				res.push_back(matrix[i][right]);
    			right--;
    		}
    		if (direction == 2) { // bottom lane, go left
    			for (int i = right; i >= left; i--) 
    				res.push_back(matrix[down][i]);
    			down--;
    		}
    		if (direction == 2) { // left lane, go top
    			for (int i = down; i >= up; i--) 
    				res.push_back(matrix[i][left]);
    			left++;
    		}
    		if (left > right || up > down)
    			break;
    		direction = (direction + 1) % 4;
    	}
    	return res;
    }
};


int main() {
	
	int mat[3][3] = {{1,2,3},{4,5,6},{7,8,9}};
	vector<vector<int> > matrix;
	for (int i = 0; i < 3; i++){
		vector<int> temp(mat[i], mat[i] + sizeof(mat[i]) / sizeof(int));
		matrix.push_back(temp);
	}
	Solution s;
	vector<int> res = s.spiralOrder(matrix);
	for (int i = 0; i < res.size(); i++)
		cout << res[i] << endl;

}

File: python/54_Spiral_Matrix.py
#!/usr/bin/python
import sys
import os


'''
[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
'''

# math trick, source http://www.cnblogs.com/zuoyuan/p/3769829.html
class Solution:
    # @param matrix, a list of lists of integers
    # @return a list of integers
    def spiralOrder(self, matrix):
        if matrix == []: return []
        up = 0; left = 0
        down = len(matrix)-1
        right = len(matrix[0])-1
        direct = 0  # 0: go right   1: go down  2: go left  3: go up
        res = []
        while True:
            if direct == 0:
                for i in range(left, right+1):
                    res.append(matrix[up][i])
                up += 1
            if direct == 1:
                for i in range(up, down+1):
                    res.append(matrix[i][right])
                right -= 1
            if direct == 2:
                for i in range(right, left-1, -1):
                    res.append(matrix[down][i])
                down -= 1
            if direct == 3:
                for i in range(down, up-1, -1):
                    res.append(matrix[i][left])
                left += 1
            if up > down or left > right: return res
            direct = (direct+1) % 4



def main():
    s = Solution()

if __name__ == "__main__":
    main()
